ipc/provider/src/manager/evm/manager.rs

```diff
@@ -116,31 +116,42 @@ impl TopDownFinalityQuery for EthSubnetManager {
             return Err(anyhow!("invalid block hash len"));
         }
 
-        let route = subnet_id_to_evm_addresses(subnet_id)?;
-        log::debug!("getting top down messages for route: {route:?}");
-
-        let subnet_id = gateway_getter_facet::SubnetID {
-            root: subnet_id.root_id(),
-            route,
-        };
-        let gateway_contract = gateway_getter_facet::GatewayGetterFacet::new(
+        let gateway_contract = gateway_manager_facet::GatewayManagerFacet::new(
             self.ipc_contract_info.gateway_addr,
             Arc::new(self.ipc_contract_info.provider.clone()),
         );
 
-        let call = gateway_contract
-            .get_top_down_msgs(subnet_id, U256::from(epoch))
-            .block(BlockId::from(H256::from_slice(block_hash)));
-        let raw_msgs = call
-            .call()
-            .await
-            .map_err(|e| anyhow!("cannot get evm top down messages: {e:}"))?;
+        let topic1 = contract_address_from_subnet(subnet_id)?;
+        log::debug!(
+            "getting top down messages for subnet: {:?} with topic 1: {}, block hash equals: {}",
+            subnet_id,
+            topic1,
+            hex::encode(&block_hash),
+        );
 
-        let mut msgs = vec![];
-        for c in raw_msgs {
-            msgs.push(ipc_sdk::cross::CrossMsg::try_from(c)?);
+        let ev = gateway_contract
+            .event::<lib_gateway::NewTopDownMessageFilter>()
+            .from_block(epoch as u64)
+            .to_block(epoch as u64)
+            .topic1(topic1);
+
+        let mut messages = vec![];
+        for (event, meta) in query_with_meta(ev, gateway_contract.client()).await? {
+            let msg = CrossMsg::try_from(event.message)?;
+            log::debug!("received message: {:?} and meta: {:?}", msg, meta);
+
+            if meta.block_hash.0 != block_hash {
```

Comment:
It shouldn't be the case that a block hash is different for one of the events of the block, right? So if the block has a different hash it should error in the first iteration. 

Also, we should de-duplicate events here in case there are reorgs. Remember the issue we found when there was a reorg and we were fetching duplicate events with the `removed` flag.  I'll try to find the PR where Akosh addresses this.

Reply:
I think it's checking the first iteration as well? Probably because I did the parsing first (so that I can log whatever before the error), that gave the wrong impression?
Akosh did it in the `query_with_meta` function, so it should be save here.

Reply:
right, you are using `query_with_meta` not `ev.query_with_meta`.